[draft] Add a dev container and user guide

toumorokoshi opened the pull request:
It has been a common friction point for contributors (at Google and beyond) to bootstrap the development environment. In particular the setup of the ruby installation.

This change adds a new container (magic-modules-dev) with some instructions on how to use it locally. As part of this change, the container magic-modules-dev will have to be made public after being published.

Details:

- Podman was chosen over Docker as it tends to be friendly to the Googler development environment.
- Magic-modules-dev is built off of the downstream-builders to chain to a build that needs to be maintained regardless.
- The readme was modified to remove commands that are difficult to run in a container and were also unsused to minimize confusion for container users.

<!-- Put a description of what this PR is for here, along with any references to issues that this resolves or contributes to -->




<!--
Replace each [ ] with [X] to check it. Switch to the preview view to make it easier to click on links.
These steps will speed up the review process, and we appreciate you spending time on them before sending
your code to be reviewed.
-->
If this PR is for Terraform, I acknowledge that I have:

- [ ] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [ ] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-the-terraform-providers), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [ ] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [ ] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [ ] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

<!-- AUTOCHANGELOG for Downstream PRs.

Please select one of the following "release-note:" headings:
    - release-note:enhancement
    - release-note:bug
    - release-note:note
    - release-note:new-resource
    - release-note:new-datasource
    - release-note:deprecation
    - release-note:breaking-change
    - release-note:none
    
Unless you choose release-note:none, please add a release note.

See .ci/RELEASE_NOTES_GUIDE.md for writing good release notes.

You can add more release note blocks if you want more than one CHANGELOG
entry for this PR.
-->
**Release Note Template for Downstream PRs (will be copied)**

```release-note:REPLACEME

```

modular-magician commented:
Oops! It looks like you're using an unknown release-note type in your changelog entries:
* REPLACEME

Please only use the types listed in https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md.

ScottSuarez commented on .ci/containers/magic-modules-dev/Dockerfile:
```diff
@@ -0,0 +1,3 @@
+FROM gcr.io/graphite-docker-images/downstream-builder:latest
+WORKDIR /magic-modules
+ENTRYPOINT ["/usr/bin/make"]
```
The container spends time fetching dependencies for tpgtools. 
This is duplicate work that can be avoided.

We'll likely have to change downstream builder to fetch these dependencies and update the trigger to work off the go.sum file.

toumorokoshi replied:
good call! I'll take a quick look, but worst case would a follow-up PR work?

toumorokoshi replied:
I think follow-up in a separate issue?

ScottSuarez reviewed (commented):
This works and is really neat. Comments surrounding messaging and implementation details.

toumorokoshi commented on README.md:
````diff
@@ -15,23 +15,93 @@ change.
+
+Here is an example of how to build Terraform (after [cloning the provider](#cloning-terraform-providers)):
+
+```shell
+podman run --rm \
+  -v $PWD:/magic-modules -v $GOPATH:$GOPATH \
+  -it gcr.io/graphite-docker-images/magic-modules-dev:latest \
+  terraform VERSION=ga \
+  OUTPUT_PATH="$GOPATH/src/github.com/hashicorp/terraform-provider-google"
+```
+
+Generally, you can replace any reference to `make` in this guide with the following:
+
+```shell
+podman run --rm \
+  -v $PWD:/magic-modules -v $GOPATH:$GOPATH \
+  -it gcr.io/graphite-docker-images/magic-modules-dev:latest
+```
````
good call! I think the command you approached makes sense. 

There's a little bit of trickiness with the ability to select podman vs docker, for example, but that could be a parameter.

ScottSuarez replied:
to figure out if it's present on the system you could run 

some pseudo code 
```bash
hasPodman=$(which podman)
hasDocker=$(which docker)

# if has podman
   containerEnvironment=podman
# else if has docker
   containerEnvironment=docker
# else
# -- error? build local? 
# end

$containerEnvironment run --rm  ### ect...
```

toumorokoshi replied:
yeah, I think this has to be a shell script to allow this type of conditioning. Also `make` doesn't lend itself well to pass in positional arguments (it's all keyword based).

I'll take this suggestion but express it as a "make wrapper" of a shell script.

toumorokoshi commented on README.md:
```diff
@@ -48,55 +118,15 @@ To get started, you'll need:
     * For M1 Mac users, run `RUBY_CFLAGS="-Wno-error=implicit-function-declaration" rbenv install 2.6.0`
 * [`Bundler`](https://github.com/bundler/bundler)
   * This can be installed with `gem install bundler`
+* Gems for magic-modules
+  * This can be installed with `cd mmv1 && bundler install`
```
this used to be in the bootstrap script, but the coupled the cloning of the TPG repos and the bundle install (now handled by the container) together.

I don't think this is anywhere else in the readme or scripts anymore, but I'll do a scanthrough before resolving.

toumorokoshi commented on README.md:
```diff
@@ -15,23 +15,93 @@ change.
+podman run --rm \
+  -v $PWD:/magic-modules -v $GOPATH:$GOPATH \
+  -it gcr.io/graphite-docker-images/magic-modules-dev:latest \
+  terraform VERSION=ga \
```
The container just wraps "make", which makes it a drop-in replacement for any mention of the string `make` in the readme (it's in a few places).

I think this is better than, for example, forcing `make terraform`: this approach prevents bundling in dependencies and running other tools like tpgtools or terraform-validator.

toumorokoshi commented on README.md:
```diff
@@ -15,23 +15,93 @@ change.
+you're encountering issues with vendoring and paths, make sure both MM and the
+Terraform provider are running on up to date copies of `main`.
+
+### Setting up a container-based environment
```
under the hood both docker and podman use Linux cgroups (https://man7.org/linux/man-pages/man7/cgroups.7.html) and namespaces (https://en.wikipedia.org/wiki/Linux_namespaces) to isolate processes that run. It also mounts the filesystem using a filesystem of choice (e.g. overlayfs).

There is a very low overhead for these tools: they're built into the Linux Kernel and for the most part (besides some syscall overhead) run the same instructions.

AFAIK Performance limitations would only be introduced by:

1. usage limits on memory and CPU for cgroups. IIRC docker and podman don't limit this by default (e.g. https://docs.podman.io/en/latest/markdown/podman-run.1.html#cpu-quota-limit) 
2. performance overhead due to I/O is the other one. pure mounts should have no difference. writes to the docker filesystem (typically using overlayfs) might have a performance difference, but IIRC it's an in-memory file system for changes so shouldn't have an impact.

I'll test for fun, but the overhead if done wrote is basically undetectable.

toumorokoshi replied:
oh, one note is that for OSX likely these *will* run slower, due to the fact that the only way you can get containers running on OSX is via virtualization of a Linux operations system (e.g. via QEMU). 

I don't have access to a OSX machine, but maybe you could test with time to see the perf difference? even for VMs these days due to VT-x the perf difference shouldn't be that different.

toumorokoshi commented on README.md:
````diff
@@ -15,23 +15,93 @@ change.
+Or build it yourself:
+
+```shell
+podman build --format=docker \
+    .ci/containers/downstream-builder/ \
+    -t gcr.io/graphite-docker-images/downstream-builder:latest
+podman build --format=docker \
+    .ci/containers/magic-modules-dev/ \
+    -t gcr.io/graphite-docker-images/magic-modules-dev:latest
````
you don't really need container local: you can try to pull and if it fails, then you fallback to build.

But yes, will add.

toumorokoshi commented on README.md:
````diff
@@ -15,23 +15,93 @@ change.
+  -it gcr.io/graphite-docker-images/magic-modules-dev:latest
+```
+
+### Preparing your environment manually
````
should I move containers to the bottom? 

My concern with moving the theoretically easier method to the bottom is that people will do the harder instructions first.

ScottSuarez replied:
We can add a follow up issue to add the make file and recommend this as the default way to build later?

toumorokoshi replied:
done as we now use a shell script.

ScottSuarez reviewed (approved):
Approving, opinions are non-blocking

toumorokoshi commented on README.md:
```diff
@@ -15,23 +15,93 @@ change.
+### Preparing your environment manually
+
+It is also possible to set up the magic-modules tooling
+on your local development environment.
```
added to the new version!

modular-magician commented:
Oops! It looks like you're using an unknown release-note type in your changelog entries:
* REPLACEME

Please only use the types listed in https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md.

toumorokoshi commented:
@ScottSuarez addressed your feedback to create a wrapper script. make-in-make doesn't provide a good UX (hard to choose directives or pass-through arguments), so I went with a shell script. PTAL.

toumorokoshi commented on scripts/make-in-container.sh:
```diff
@@ -0,0 +1,49 @@
+    echo "NOTE: this may take a while."
+    sleep 1
+    if ! ${CONTAINER_EXECUTABLE} images | grep "${BUILDER_IMAGE}" > /dev/null; then
+        if ! ${CONTAINER_EXECUTABLE} pull "${BUILDER_IMAGE}"; then
```
this will fail if someone doesn't have access or the container doesn't exist, so it works for both those who have access (Googlers) and for third parties (it'll just move to building instead).

ScottSuarez replied:
very cool ^^

ScottSuarez commented:
Question, couldn't we call this script from make? I'm not sure the limitations there but we should be able to run scripts and persist data, correct? Not saying we have to. Also, might be nice to consolidate this into the tools folder unless you disagree.

toumorokoshi commented:
> Question, couldn't we call this script from make? I'm not sure the limitations there but we should be able to run scripts and persist data, correct? Not saying we have to. Also, might be nice to consolidate this into the tools folder unless you disagree.

Can you sketch how this would work? I could imagine a couple approaches, neither of which are better than the current shell script wrapper IMO:

1. you make a container variant of each make command, which takes the same arguments. I avoided that because it's a lot of duplication (increase in maintenance cost) at minimal benefit over a shell command.

2. you can create a virtual args argument and use that (example: https://stackoverflow.com/questions/2214575/passing-arguments-to-make-run). This deviates slightly from the syntax and examples of all the other make invocations.

There is no way to pipe positional arguments directly into make directives, without some hacky magic: https://pawamoy.github.io/posts/pass-makefile-args-as-typed-in-command-line/.

ScottSuarez commented:
Oh I see what you're saying with recursive make. The only alternative I could see would be executing the CLI commands directly against the container when container mode is enabled instead of having make be the entrypoint.

toumorokoshi commented:
> Oh I see what you're saying with recursive make. The only alternative I could see would be executing the CLI commands directly against the container instead of having make be the entrypoint. I think you mention this though with the command variants.

yeah it's tricky. I think we should try this and if we get feedback that a different way of doing it is preferred, we can always update.

I know you gave an LGTM already, but any other concerns before I merge?

ScottSuarez commented:
No no concerns that I can think of

toumorokoshi commented:
SGTM! merging then.
